README.md
# CarrIOTA Nelson GUI/Monitor

This is a small http server that monitors the Nelson instance and outputs the status as a web page.
Based on [create-react-app](https://github.com/facebookincubator/create-react-app).

## Getting Started

### Prerequisites

- A Nelson instance (including IRI) should be running locally on your machine.
- `node` (version >6.9) and `yarn` and `npm` should be installed
- install `serve` globally: `sudo npm install -g serve`
- Use Firefox or Chrome browsers; there may be problems with IE (specially everything before Edge)

### Installing

Globally install Nelson

```
npm install -g @semkodev/nelson.gui
```

And run it

```
nelson.gui -p 5000
```

### Running as a service

You can use the [node process manager](http://pm2.keymetrics.io/) to run Nelson GUI as a service.
Just do the following:

```
# Install the process manager:
npm install pm2 -g

# Make pm2 start at startup:
pm2 startup

# Start the Nelson GUI as service
pm2 start nelson.gui --port 3333

# Save current processes runing with pm2 to startup on boot:
pm2 save
```

### Docker

Provided you have docker installed, Nelson GUI can be started as follows:

```
docker run <docker opts> semkodev/nelson.gui <nelson.gui command line opts>
```

### Installing Manually

Clone the project

```
git clone git@gitlab.com:semkodev/nelson.gui.git
cd nelson.gui
yarn install --pure-lockfile
node build/nelson.gui.js
```

Now your server runs at `http://localhost:5000`

### Connecting to Nelson on another host/port

Is your Nelson's API running on a non-standard host:port (localhost:18600)?
Simply provide it when starting the GUI:

```
nelson.gui -p 5000 --apiHostname another.hostname.com --apiPort 123456
```

The data is refreshed automatically each 5 seconds (or at the heartbeat rate of your Nelson)

### Connecting to Nelson +v.0.4.0 with username/password

If your Nelson's API is username/password protected (strongly recommended!),
you can specify the credentials in the URL:

```
http://www.your.gui.host.com:<port>/#/<username>:<password>
```

This is the simplest way. A better login window is coming soon.

**NOTE**: If your Nelson API is not password protected, the GUI will
scramble all the hostnames and IPs for security reasons.

## Development

The monitor is under heavy development. Please be patient.
Contribution / Pull Requests are welcome, but at this early stages where so much changes, your
work might become quickly non-mergeable and thus be in vain.
Please, for now, refrain to small PRs, thank you!

### Donations

**Donations always welcome**:

```
AAJXXFJUEQHKPYIOUIUO9FWCMOAFBZAZPXIFRI9FLDQZJGHQENG9HNMODUZJCHR9RHHUSBHWJELGRDOWZRNWYLYCQW
```

### What's Next

- Make it responsive
- Add optional listing of all known hosts (already fetching)
- Add a dialog to change hostname/port of Nelson
- Implement Menu
- Add some visualization of nodes / connections
- Add favicon / logo and some visuals
- Add version of the package to the Web output
- Add tests
- Improve documentation & clean up code
- ...


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.5] - 2018-06-11

### Changes

- Chnages package signature

## [0.2.3] - 2018-05-08

### Changes

- Fixes update interval and adds sync velocity stats

## [0.2.3] - 2018-04-12

### Changed

- Fixed data not loading in the first 5 seconds.

## [0.2.2] - 2018-02-04

### Changed

- Fixed api hostname and logging.

## [0.2.1] - 2018-02-04

### Changed

- Fixed api requests overflooding.

## [0.2.0] - 2018-02-04

### BREAKING CHANGE

- The url structure has changed. Please read the README!

### Added

- Support for Nelson 0.4.0
- Proxy to a Nelson server (you do not have to specify it in the URL any more)
- More information about the running IRI instance
- Stats about new/connected nodes

### Changed

- Improved README
- Hide IPs if the Nelson API is not password-protected


src/nelson.gui.js
#!/usr/bin/env node
const request = require('request');
const md5 = require('md5');
const crypto = require('crypto');
const express = require('express');
const throttle = require("express-throttle");
const bodyParser = require('body-parser');
const program = require('commander');
const path = require('path');
const app = express();
const version = require('../package.json').version;

const apiThrottleOptions = {
    "burst": 5,
    "rate": "10/min"
};

const parseNumber = (v) => parseInt(v);

process.on('unhandledRejection', (reason, p) => {
    console.log('Unhandled Rejection at: Promise', p, 'reason:', reason);
});

program
    .version(version)
    .option('-p, --port [value]', 'Nelson GUI port', parseNumber, 5000)
    .option('-a, --apiPort [value]', 'Nelson API port', parseNumber, 18600)
    .option('-h, --apiHostname [value]', 'Nelson API hostname', 'localhost')
    .parse(process.argv);

// Define the port to run on
app.set('port', process.env.NELSON_GUI_PORT || program.port);

app.use(express.static(path.join(__dirname)));
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());
app.get('/api', throttle(apiThrottleOptions), proxy);
app.get('/api/peers', throttle(apiThrottleOptions), proxy);

const salt = crypto.randomBytes(48).toString('hex');
// Listen for requests
const server = app.listen(app.get('port'), function() {
    console.log(`Nelson GUI running on port ${server.address().port}`);
});

const scramble = (str) => md5(`${salt}:${str}`).substr(0, 6);

function scramblePeer(peer) {
    return {
        ...peer,
        ip: scramble(peer.ip),
        hostname: scramble(peer.hostname)
    }
}

function proxy (req, resp) {
    const path = req.originalUrl.replace('/api', '');
    const url = `http://${program.apiHostname}:${program.apiPort}${path}`;
    const opts = { url };
    if (req.query.auth) {
        const tokens = req.query.auth.split(':');
        opts.auth = { user: tokens[0], pass: tokens[1] }
    }
    request(opts, (err, r, body) => {
        if (err) {
            console.log(`Unexpected error happened: ${err}`);
            resp.status(500);
            return;
        }
        if (r.statusCode === 401) {
            console.log('Remote access denied');
            resp.status(401);
            return;
        }
        try {
            const json = JSON.parse(body);
            let response = json;
            if (path.startsWith('peers') && !req.query.auth) {
                response = json.map(scramblePeer);
            } else if(response.connectedPeers && !req.query.auth) {
                response.connectedPeers = response.connectedPeers.map(scramblePeer)
            }
            resp.status(200).json(response);
        }
        catch (e) {
            console.log(`Remote returned unexpected response(${r.statusCode})`, e);
            resp.status(500);
        }
    })
}